e2e-tests/src/ban.rs

```diff
@@ -124,3 +138,58 @@ pub fn check_ban_event<C: AnyConnection>(
 
     Ok(event)
 }
+
+pub fn check_session_count<C: AnyConnection>(
+    connection: &C,
+    seats: &CommitteeSeats,
+    reserved_validators: &[AccountId],
+    non_reserved_validators: &[AccountId],
+    start_session: &SessionIndex,
+    sessions_to_check: &SessionCount,
+    ban_session_threshold: &SessionCount,
+) -> anyhow::Result<()> {
+    let validators: Vec<_> = reserved_validators
+        .iter()
+        .chain(non_reserved_validators.iter())
+        .collect();
+
+    let mut expected_validator_session_count = HashMap::new();
+
+    for session in *start_session..*start_session + *sessions_to_check {
+        wait_for_at_least_session(connection, session)?;
+
+        let reserved_members_for_session =
+            get_members_subset_for_session(seats.reserved_seats, reserved_validators, session - 1);
+        let non_reserved_members_for_session = get_members_subset_for_session(
+            seats.non_reserved_seats,
+            non_reserved_validators,
+            session - 1,
+        );
+        let members_for_session: Vec<_> = reserved_members_for_session
+            .iter()
+            .chain(non_reserved_members_for_session.iter())
+            .collect();
+
+        validators.iter().for_each(|&val| {
+            info!(
+                "Checking session count | session {} | validator {}",
+                session - 1,
+                val
+            );
+            let session_count = expected_validator_session_count.entry(val).or_insert(0);
+
+            if members_for_session.contains(&val) {
+                *session_count += 1;
+                *session_count %= ban_session_threshold;
+            }
+
+            let expected_session_count = expected_validator_session_count
+                .get(&val)
+                .unwrap_or_else(|| panic!("Missing expected session count for validator {}", val));
+
+            check_underperformed_validator_session_count(connection, val, expected_session_count);
```

Comment:
from task/test description it should be enough to just check if the validator that was in the committee was marked as underperformer - that would get rid of maps, modulos and counter greatly simplifying the test

Reply:
This is also changed. There is still some less-than-pretty conditional wrangling since the diff in counts between sessions depends on whether the particular validator is in committee and, additionally, the counter might be reset.